product/scanning.py

```diff
@@ -116,40 +116,172 @@ def _retrieve_cached_scan_result(
     ):
         return self.retrieve_scan_result(container_image=container_image, component=component)
 
-    def upload_image(
-            self,
-            container_image: ContainerImage,
-            component: Component,
-            wait_for_result: bool=False
-        ):
-        metadata = self._metadata(container_image=container_image, component=component)
+    def process_image(
+        self,
+        container_image: ContainerImage,
+        component: Component,
+        processing_mode: ProcessingMode,
+        wait_for_result: bool=False,
+    ):
+        if processing_mode is ProcessingMode.FORCE_UPLOAD:
+            # Shortcut: Upload image no matter what
+            default_custom_data = self._metadata(
+                container_image=container_image,
+                component=component,
+            )
+            processing_result = self.upload_image_to_protecode(
+                container_image=container_image,
+                component=component,
+                custom_data=default_custom_data,
+            )
 
-        upload_result = partial(UploadResult, container_image=container_image, component=component)
+        elif processing_mode is ProcessingMode.UPLOAD_IF_CHANGED:
+            processing_result = self.upload_image(
+                container_image=container_image,
+                component=component,
+            )
 
-        # check if the image has already been uploaded for this component
-        scan_result = self.retrieve_scan_result(
+        elif processing_mode is ProcessingMode.RESCAN:
+            processing_result = self.rescan_image(
+                container_image=container_image,
+                component=component,
+            )
+
+        else:
+            raise NotImplementedError
+
+        return self.wait_for_result(
+            upload_result=processing_result,
+            wait_for_result=wait_for_result,
+        )
+
+    def wait_for_result(
```

Comment:
the wait_for_result method seems like a good refactoring to me. however: why does it have a wait_for_result argument? what's the point of calling wait_for_result with this argument set to false? in that case, it would not make sense to call this method at all, would it? unless it did something else. which would indicate that:
a) it should be named differently
b) the part that does other things should be split into a separate method